packages/gatsby/src/utils/webpack.config.js

```diff
@@ -332,6 +333,11 @@ module.exports = async (
         ),
         "create-react-context": directoryPath(`.cache/create-react-context.js`),
       },
+      plugins: [
+        PnpWebpackPlugin.bind(directoryPath(`.cache`), module),
+        PnpWebpackPlugin.bind(directoryPath(`public`), module),
+        PnpWebpackPlugin,
```

Comment:
What does this instance do?

Reply:
With PnP one important thing is that a file is only allowed to require the third-party that its package depends on. So for example you can't accidentally `require('lodash')` if you forgot to list it in your dependencies (which would work otherwise but could break in production if the lodash you were relying on was obtained via a dev dependency).

In order to do this, we rely on the path of the module that makes the require call. Since Yarn knows the location of all the packages in your dependency tree, we can safely deduce that if a package is installed in `/X/Y`, then a file located in `/X/Y/Z.js` can only require the files listed in `/X/Y/package.json`.

Unfortunately Gatsby breaks this model by creating files within the user folders. Those files are copied from the `gatsby` package (gonna be honest I'm still not entirely sure why they are copied in the first place rather than used from their original location), and put inside the user folder. What this mean is that when `/X/Y/.cache/default-html.js` makes a require to something, Yarn thinks that the dependency should be resolved relative to `/X/Y/package.json` - even though it should instead be from something like `/cache/yarn/gatsby-1.2.3/package.json`.

The `bind` directive instruct the Webpack PnP plugin that everything in the specified directory path should be resolved as if it was the specified module object. Since `src/utils/webpack.config.js` is kept within the Gatsby package folder, the cache files require calls are resolved relative to the dependencies listed in it and everything works.